### DISCUSSION TITLE
Spot Simulation

### DISCUSSION THREAD

**Discussion by kahleeeb3** [bdaiinstitute/spot_ros2#109] [Q&A] [answered]
Is there any sort of simulation program for SPOT where I can test my code when I do not have the robot close by?

**Comment by METEORITENMAX** [comment, marked as answer]
Hi Caleb,
yes we have a sort of simulation for Spot to test our software. Unfortunately, at the moment the simulation only allows us to test software that does not interfere with the Spot SDK. So we for example we test ROS2 related stuff like Navigation2, frontier and our high-level logic language: golog++ in simulation before running on Spot.

If you want to try out, maybe first have a look into our web approach:
https://github.com/RRL-ALeRT/spot-webots-cloud

And here you find the ros2 package for spot in webots:
https://github.com/MASKOR/webots_ros2_spot

Greetings,
Max

**Reply by kahleeeb3** [reply]
Hi @METEORITENMAX, thanks for the response. I'm a little confused about this simulation. Once I have it running how do I connect to the virtual SPOT in order to run the example code from this repository? i.e. How would I do this for the simulation?
```
ros2 launch spot_driver spot_driver.launch.py
ros2 run simple_walk_forward walk_forward
```

**Reply by METEORITENMAX** [reply]
Hi Caleb,
unfortunately, the web solution does not work at the moment..

However, if you followed the instructions on the webots_ros2_spot (https://github.com/MASKOR/webots_ros2_spot) repo, installed and sourced everything, you should start the spot webots world via:
```
ros2 launch webots_spot spot_launch.py
```
This will launch a webots world with Spot inside. With that `spot_launch.py` several ROS2 topics and services are started.
To see all, execute in a sourced terminal:
```
ros2 topic list
ros2 service list
```
Or launch `rqt` in a terminal to see and use them.
E.g. after you launch `spot_launch.py` a ROS2 topic `\cmd_vel` is available and when you launch:
```
ros2 run spot_teleop spot_teleop_keyboard
```
or the default:
```
ros2 run teleop_twist_keyboard teleop_twist_keyboard
```
you can move Spot with your keyboard.

The `spot_launch.py` starts nodes that provide an interface to the virtual Spot by ROS2 topics, services and actions.
Inside the README there are also the commands to launch MoveIt for controlling the arm and Navigation2.

I hope this helps you. If you have any further questions, feel free to ask.

Greetings,
Max

**Reply by kahleeeb3** [reply]
Thanks Max, that was very helpful. I found that running
```
ros2 run spot_teleop spot_teleop_keyboard
```
resulted in an error:
```
Package 'spot_teleop' not found
```
Using the 
```
ros2 run teleop_twist_keyboard teleop_twist_keyboard
```
Command did work, however, I could not get spot to walk forward without falling over.
